rest/data/commit_queue.go

```diff
@@ -69,19 +73,90 @@ func (pc *DBCommitQueueConnector) FindCommitQueueByID(id string) (*restModel.API
 	return apiCommitQueue, nil
 }
 
-func (pc *DBCommitQueueConnector) CommitQueueRemoveItem(id, item string) (bool, error) {
+func (pc *DBCommitQueueConnector) CommitQueueRemoveItem(id, item string) (*commitqueue.CommitQueueItem, error) {
 	cq, err := commitqueue.FindOneId(id)
 	if err != nil {
-		return false, errors.Wrapf(err, "can't get commit queue for id '%s'", id)
+		return nil, errors.Wrapf(err, "can't get commit queue for id '%s'", id)
 	}
 
-	return cq.Remove(item)
+	head := cq.Next()
+	itemRemoved, err := cq.Remove(item)
+	if err != nil {
+		return itemRemoved, errors.Wrapf(err, "can't remove item '%s' from queue '%s'", item, id)
+	}
+
+	if itemRemoved != nil && head.Issue == item {
+		if err = preventMergeForItem(id, head); err != nil {
+			return itemRemoved, errors.Wrapf(err, "can't prevent merge for item '%s' on queue '%s'", item, id)
+		}
+	}
+
+	return itemRemoved, nil
 }
 
 func (pc *DBCommitQueueConnector) CommitQueueClearAll() (int, error) {
 	return commitqueue.ClearAllCommitQueues()
 }
 
+func preventMergeForItem(projectID string, item *commitqueue.CommitQueueItem) error {
+	projectRef, err := model.FindOneProjectRef(projectID)
+	if err != nil {
+		return errors.Wrapf(err, "can't find projectRef for '%s'", projectID)
+	}
+
+	if projectRef.CommitQueue.PatchType == commitqueue.PRPatchType {
+		if item.Version != "" {
+			// Clear the subscription
+			subscriptions, err := event.FindSubscriptions(event.ResourceTypePatch, []event.Selector{{Type: event.SelectorID, Data: item.Version}})
+			if err != nil {
+				return errors.Wrapf(err, "can't find subscription to patch '%s'", item.Version)
+			}
+			for _, subscription := range subscriptions {
+				if subscription.Subscriber.Type == event.GithubMergeSubscriberType {
+					err = event.RemoveSubscription(subscription.ID)
+					if err != nil {
+						return errors.Wrap(err, "can't remove subscription for GitHub merge")
+					}
+				}
+			}
+		}
+	}
+
+	if projectRef.CommitQueue.PatchType == commitqueue.CLIPatchType {
+		version, err := model.VersionFindOneId(item.Issue)
+		if err != nil {
+			return errors.Wrapf(err, "can't find patch '%s'", item.Issue)
+		}
+		if version != nil {
+			// Clear the subscription
+			subscriptions, err := event.FindSubscriptions(event.ResourceTypePatch, []event.Selector{{Type: event.SelectorID, Data: version.Id}})
```

Comment:
This code is exactly the same as the PRPatchType, besides Selector.Data & the subscriber type. Can this be made into a helper function that takes in selector.Data for clarity?

Reply:
done.